src/api/wallet/WalletApi.ts

```diff
@@ -0,0 +1,230 @@
+import { WalletApi, Network, WalletInfo, Command } from 'types'
+import BN from 'bn.js'
+import assert from 'assert'
+
+import WalletConnectProvider from '@walletconnect/web3-provider'
+
+import Web3 from 'web3'
+import { BlockHeader } from 'web3-eth'
+
+import {
+  getProvider,
+  getProviderState,
+  createSubscriptions,
+  Provider,
+  WalletConnectInits,
+  isWalletConnectProvider,
+  isMetamaskSubscriptions,
+  isWalletConnectSubscriptions,
+  isMetamaskProvider,
+  Subscriptions,
+} from '@gnosis.pm/dapp-ui'
+
+import { log, toBN } from 'utils'
+
+type OnChangeWalletInfo = (walletInfo: WalletInfo) => void
+
+// to track chain state, be that current account balance or token balance of that account
+// or any data on the chain belonging to that account
+// we need to refetch that data when
+// 1: network changes
+// 2: account changes
+// 3: new block is mined
+
+interface BlockchainUpdatePrompt {
+  account: string
+  chainId: number
+  blockHeader: BlockHeader | null
+}
+
+// provides subscription to blockhain updates for account/network/block
+const subscribeToBlockchainUpdate = ({
+  provider,
+  subscriptions,
+  web3,
+}: {
+  provider: Provider
+  subscriptions?: Subscriptions
+  web3: Web3
+}) => {
+  const subs = subscriptions || createSubscriptions(provider)
+
+  let networkUpdate: (callback: (chainId: number) => void) => () => void
+
+  if (isMetamaskSubscriptions(subs)) networkUpdate = cb => subs.onNetworkChanged(networkId => cb(+networkId))
```

Comment:
auto cast to number

Reply:
cute!

Reply:
I should probably do this casting inside the lib, no reason to leave it to to whoever uses the lib

Reply:
That thingy is called unary plus :smile:

Reply:
`networkId` comes from an abstraction defined by us. Why is not already typed as a number?

Also, I'm not super fan of this `+` thing, `parseInt(networkId)` is more declarative

Reply:
Because they are different subscriptions
MMask gives a string `on('networkChanged')`
WC gives number for `on('chainChanged')` but string for `on('networkChanged')`, BUT the latter one is never (almost never) fired.
So rather than invent subscriptions that don't actually exist (you can see how we detect that MMask is logged out on L177), I decided to leave usage details to the user

`parseInt` is more permissive in its detection

Reply:
I meant, yes I assume they can be different. But we did a library that abstracts away the difference, so it seems to me the right place to use the right subscription for each wallet type, and cast if necessary

Maybe my question is more related to the `'@gnosis.pm/dapp-ui'`, cause I couldn't review the PR because it was big, but what I would expect from a library like this one is to abstract away the differences instead of doing in the app code like `if isMetamaskSubscriptions(...) the n A else B`.

Probably after breaking `dapp-ui` PR and doing a proper PR I can understand better.